Add documentation for Data Shaping (#185)
osisoft/OCS-Docs: Documentation/DataViews/Data_View_Shape.md

﻿---
uid: DataViewShape
---

# Data View Shape

Data views may be set to resolve as a standard shape or a "narrow" shape. 

With a standard shape, each row in the resolved data view includes all the data fields for a single event or observation. With a narrow shape, each row in the resolved data view includes only one data field. This results in a narrow output schema where one column contains all the data field values, so the schema remains fixed regardless of changes to the included data fields. Narrow shape may be used when an invariant output schema is required.

## Specify the shape

> Set the `DataView` `Shape` property to either `DataViewShape.Standard` (default) or `DataViewShape.Narrow`.

Reviewer: What do you think about adding a link to either the DataView page, or specifically, the Shape property section in the Data View page, here?

Author: sure

#### Standard shape

> The standard table column structure is built horizontally from left to right.  The index column is first, followed by the grouping fields, if any exist.  Data field sets come next, in the order they are presented in the data view. Fields are presented in order of appearance within in each data field set for each stream from the associated query.

Reviewer: extra space after the first two periods in this paragraph.

Author: fixed

> The standard table column structure is built horizontally from left to right.  The index column is first, followed by the grouping fields, if any exist.  Data field sets come next, in the order they are presented in the data view. Fields are presented in order of appearance within in each data field set for each stream from the associated query.

Reviewer: Instead of "index column", can we use "index field"? This term is consistent with the property name of the data view.

Author: fixed

> The standard table column structure is built horizontally from left to right.  The index column is first, followed by the grouping fields, if any exist.  Data field sets come next, in the order they are presented in the data view. Fields are presented in order of appearance within in each data field set for each stream from the associated query.

Reviewer: Typo: "within in each data field set"

Author: fix

##### Example: Get data response body for standard shape with no grouping field

Note, column headers contain a numerical suffix to ensure that each column header is unique.

```csv
Timestamp.0,Id.1,SolarRadiation.2,Temperature.3,Tags.4,Id.5,SolarRadiation.6,Temperature.7,Tags.8,Id.9,SolarRadiation.10,Temperature.11,Tags.12,Id.13,SolarRadiation.14,Temperature.15,Tags.16
2018-11-13T18:19:00.0000000Z,WS_WINT,184,,"Gen2, High Resolution, Weather",WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather",WS_BILT,229,28.042327888795327,"Gen1, High Resolution, Weather",WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
2018-11-13T18:20:00.0000000Z,WS_WINT,184,,"Gen2, High Resolution, Weather",WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather",WS_BILT,229,28.042327888795327,"Gen1, High Resolution, Weather",WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
```
<br />

##### Example: Get data response body for standard shape with grouping field (Site)

> The use of a grouping field will restructure the table vertically.

Reviewer: Perhaps add a link to the "Sectioning" page.

Author: OK

2018-11-13T18:19:00.0000000Z,Hearst Castle,WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
2018-11-13T18:20:00.0000000Z,Hearst Castle,WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
2018-11-13T18:21:00.0000000Z,Hearst Castle,WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
2018-11-13T18:22:00.0000000Z,Hearst Castle,WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
2018-11-13T18:19:00.0000000Z,Rosecliff,WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather"
2018-11-13T18:20:00.0000000Z,Rosecliff,WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather"
2018-11-13T18:21:00.0000000Z,Rosecliff,WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather"
2018-11-13T18:22:00.0000000Z,Rosecliff,WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather"
2018-11-13T18:19:00.0000000Z,Winterthur,WS_WINT,184,,"Gen2, High Resolution, Weather"
2018-11-13T18:20:00.0000000Z,Winterthur,WS_WINT,184,,"Gen2, High Resolution, Weather"
2018-11-13T18:21:00.0000000Z,Winterthur,WS_WINT,184,,"Gen2, High Resolution, Weather"
2018-11-13T18:22:00.0000000Z,Winterthur,WS_WINT,184,,"Gen2, High Resolution, Weather"
```
<br />

#### Narrow shape

> The narrow table ‘pivots’ the standard table. Each data field becomes a row comprised of the following columns: 

Reviewer: Shall we use double quote around "pivots", so it is consistent with the remainder of the doc?

Author: fixed

> * Timestamp column 

Reviewer: Shall we use Index instead of Timestamp here? It's a little more generic. Since the other columns use generic term (group identifying column instead of Site column)

Author: fixed

> * Group identifying column(s)

Reviewer: Can we call this Grouping value column? Identifying field is the new term for distinguisher and we would want to avoid confusion by using the term identifying here.

Author: fixed

* Field column, which holds the resolved column label of the field

> * Value column, which holds the property value or stream related data such as name, id, tags or meta data

Reviewer: metadata is one word.

Author: fixed

The column header of the timestamp and group identifying columns may vary depending on the resolved label. The Field and Value column headers are not modifiable.

The data view is built veritically by grouping field, if present, then by field.  

> ##### Example: Get data response body for narrow shape

Reviewer: How about clarify this with "Get data response body for narrow shape with grouping field (Site)"

Author: fixed

﻿---
uid: DataViewShape
---

# Data View Shape

> Data views may be set to resolve as a standard shape or a "narrow" shape. 

Reviewer 2: ```suggestion
Data views may be set to resolve as a standard shape or a narrow shape. 
```

Author: changed

With a standard shape, each row in the resolved data view includes all the data fields for a single event or observation. With a narrow shape, each row in the resolved data view includes only one data field. This results in a narrow output schema where one column contains all the data field values, so the schema remains fixed regardless of changes to the included data fields. Narrow shape may be used when an invariant output schema is required.

> The shape concepts presented in this section apply to building all data view output formats; although the terminology used pertains to tabular output with headers (.csvh).

Reviewer: shall we specify form=csvh? (

Author: makes sense!

> The shape concepts presented in this section apply to building all data view output formats; although the terminology used pertains to tabular output with headers (form = csvh).

Reviewer 2: ```suggestion
The shape concepts presented in this section apply to building all data view output formats; although the terminology used pertains to tabular output.
```

Author: Actually Daphne add this, so I am going to keep it

Reviewer 2: Ah okay. Maybe one more small suggestion then:
```suggestion
The shape concepts presented in this section apply to building all data view output formats; although the terminology used, and examples shown, pertain to the tabular output form `format=csvh` described in [Getting Data](link to Getting Data section)
```

Author: OK - but do we need to say _described in Getting Data_?  To me it is redundant, I just put a link around headers that will take the reader to the Getting Data page and the format section.

Reviewer 3: Sure that's fine, we can go with what you've got in there now

## Specify the shape
Set the `DataView` [`Shape`](xref:DataView#DataViewShape) property to either `DataViewShape.Standard` (default) or `DataViewShape.Narrow`.

#### Standard shape

The standard table column structure is built horizontally from left to right. The index field is first, followed by the grouping fields, if any exist.  Data field sets come next, in the order they are presented in the data view. Fields are presented in order of appearance within each data field set for each stream from the associated query.

> Vertically, the standard structure depends on the inclusion of grouping fields. If grouping fields are not defined, each resultant index appears only once, and all interpolated data is in that row. If grouping fields are defined, then the resultant indexes will repeat vertically for each group. The groups are presented in alphabetical order.

Reviewer 2: This means alphabetical order of the grouping fields' values, right?

Author: yes

##### Example: Get data response body for standard shape with no grouping field

> Note, column headers contain a numerical suffix to ensure that each column header is unique.

Reviewer: Can we add a slight explanation why the numerical suffix was added/how to fix (e.g. because the data view did not contain enough information (via grouping/identifying field) to differentiate the different column labels).

Author: sure

```csv
Timestamp.0,Id.1,SolarRadiation.2,Temperature.3,Tags.4,Id.5,SolarRadiation.6,Temperature.7,Tags.8,Id.9,SolarRadiation.10,Temperature.11,Tags.12,Id.13,SolarRadiation.14,Temperature.15,Tags.16
2018-11-13T18:19:00.0000000Z,WS_WINT,184,,"Gen2, High Resolution, Weather",WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather",WS_BILT,229,28.042327888795327,"Gen1, High Resolution, Weather",WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
2018-11-13T18:20:00.0000000Z,WS_WINT,184,,"Gen2, High Resolution, Weather",WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather",WS_BILT,229,28.042327888795327,"Gen1, High Resolution, Weather",WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
```
<br />

##### Example: Get data response body for standard shape with grouping field (Site)

> The use of a grouping field will restructure the table vertically. See [Grouping documentation](xref:DataViewsSectioning) for details.

Reviewer 2: just double checking this doc link will start to work when its merged? In preview mode, it isn't showing "Grouping documentation" as a link

Author: yes

2018-11-13T18:19:00.0000000Z,Hearst Castle,WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
2018-11-13T18:20:00.0000000Z,Hearst Castle,WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
2018-11-13T18:21:00.0000000Z,Hearst Castle,WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
2018-11-13T18:22:00.0000000Z,Hearst Castle,WS_HRST,254,4.219790459247207,"Gen1, Low Resolution, Weather"
2018-11-13T18:19:00.0000000Z,Rosecliff,WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather"
2018-11-13T18:20:00.0000000Z,Rosecliff,WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather"
2018-11-13T18:21:00.0000000Z,Rosecliff,WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather"
2018-11-13T18:22:00.0000000Z,Rosecliff,WS_ROSE,31,31.23228568408279,"Gen1, Low Resolution, Weather"
2018-11-13T18:19:00.0000000Z,Winterthur,WS_WINT,184,,"Gen2, High Resolution, Weather"
2018-11-13T18:20:00.0000000Z,Winterthur,WS_WINT,184,,"Gen2, High Resolution, Weather"
2018-11-13T18:21:00.0000000Z,Winterthur,WS_WINT,184,,"Gen2, High Resolution, Weather"
2018-11-13T18:22:00.0000000Z,Winterthur,WS_WINT,184,,"Gen2, High Resolution, Weather"
```
<br />

#### Narrow shape

> The narrow table "pivots" the standard table. Each data field becomes a row comprised of the following columns: 

Reviewer 2: ```suggestion
The narrow table pivots the standard table. Each data field becomes a row comprised of the following columns: 
```

Author: changed
